docs/style-spec/expressions.md

```diff
@@ -0,0 +1,196 @@
+  - Literal: `null`
+- `String`
+  - Literal: `"my string value"`
+- `Number`
+  - Literal: `1729`
+- `Boolean`
+  - Literal: `true` or `false`
+- `Color`
+- `Object`
+  - Literal: `["literal", { "key": value, "key2": value2 }]`
+- `Array<T, N>`: an array of N values of type T
+  - Literal: `["literal", [v0, v1, ...]]`
+- `Array<T>`: a dynamically sized array of values of type `T`
+- `Array`: equivalent to `Array<Value>`
+- `Value`: A "variant" type representing the set of possible values retrievable from a feature's `properties` object (`Null | String | Number | Boolean | Object | Array`)
+- `Error`: a subtype of all other types. Used wherever an expression is unable to return the appropriate supertype. Carries diagnostic information such as an explanatory message and the expression location where the Error value was generated.
+
+### Constants:
+- `[ "ln2" ] -> Number`
+- `[ "pi" ] -> Number`
+- `[ "e" ] -> Number`
+
+### Variable binding:
+- `[ "let", name1: String, e1, name2: String, e2, ..., e_result ]` - Bind expression `e1` to the string `name1`, `e2` to `name2`, etc., before evaluating `e_result`.  The bound expressions may be referenced within `e_result` with `[ "var", name1 ]`, `[ "var", name2 ]`, etc. (E.g.: `["let", "a", 1, "b", ["number", ["get", "blah"]], [ "+", ["var", "a"], ["var", "b"] ]`.)
+
+### Type assertion:
+Assert that the argument is of a specific type, producing a runtime error if it is not.
+
+- `["string", e: Value] -> String`
+- `["number", e:Value] -> Number`
+- `["boolean", e:Value] -> Boolean`
+- `["object", e:Value] -> Object`
+- `["array", e:Value] -> Array`
+- `["array", T, length, e: Value] -> Array<T, length>`
+
+### Type conversion:
+Convert the argument to the given type, producing a runtime error if the conversion is not possible.
+
+- `["to_string", e: Value] -> String`
+- `["to_number", e:Value] -> Number`
+  - Uses platform-default string-to-number conversion. (TBD: parse locale-specificformatted number strings)
+- `["to_boolean", e:Value] -> Boolean`
+  - `0`, `''`, and `null` are considered falsy.
+
+### Lookup:
+- `["get", key: String, obj: Object = ["properties"] ] -> Value` - note that the final argument defaults to `["properties"]`, so to get feature property `"x"`, simply use `["get", "x"]`.
+- `["has", key: String, obj: Object = ["properties"] ] -> Boolean`
+- `["at", index: Number, arr: Array<T, N>|Array<T>] -> T`
+- `["typeof", expr: Value] -> String`
+- `["length", e: Vector<T>|String] -> Number`
+
+### Feature data:
+- `["properties"] -> Object` the feature's `properties` object
+- `["geometry_type"] -> String` the string value of `feature.geometry.type`
+- `[ "id" ] -> Value` returns the value of `feature.id`.
+
+### Decision:
+- `["case", cond1: Boolean, result1: T, cond2: Boolean, result2: T, ..., cond_m, result_m: T, result_otherwise: T] -> T`
+- `["match", x: T, a_1: T, y_1: U, a_2: T, y_2: U, ..., a_m: T, y_m: U, y_else: U]` - `a_1`, `a_2`, ... must be _literal_ values of type `T`.
```

Comment:
@anandthakker Do you think it makes sense to consider making `y_else: U` optional? Could the renderer maybe treat `y_1` (or maybe `y_m`) as this value if it's not explicitly assigned? A common usage of "match" in our core styles is using it to assign style values across `a` values that have already been explicitly filtered for in that style layer, so we know that those are the only possible `a` values. Including an `y_else: U` value in all of these expressions is fine, but seems a bit inelegant, since they contain values that will never be used.